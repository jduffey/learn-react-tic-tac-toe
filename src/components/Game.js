import React from 'react';

import { Board } from "./Board";
import { calculateWinner } from '../utils/calculateWinner';

const [playerOneMark, playerTwoMark] = ['X', 'O'];

export class Game extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            history: [{
                squares: Array(9).fill(null),
            }],
            stepNumber: 0,
            isPlayerOneNext: true,
        };
    }

    handleClick(i) {
        const history = this.state.history;
        const current = history[history.length - 1];
        const squares = current.squares.slice(); // Make a copy of the data instead of mutating it

        if (calculateWinner(squares).winner || squares[i]) {
            return;
        }

        squares[i] = this.state.isPlayerOneNext ? playerOneMark : playerTwoMark;
        this.setState({
            history: history.concat([{
                squares: squares,
            }]),
            stepNumber: history.length,
            isPlayerOneNext: !this.state.isPlayerOneNext,
        });
    }

    jumpTo(step) {
        this.setState({
            stepNumber: step,
            isPlayerOneNext: (step % 2) === 0,
        });
    }

    render() {
        const history = this.state.history;
        const current = history[this.state.stepNumber];
        const winningData = calculateWinner(current.squares);

        const moves = history.map((_, move) => {
            const desc = move ?
                'Go to move #' + move :
                'Go to game start';
            return (
                <li key={move}>
                    <button onClick={() => this.jumpTo(move)}>
                        {desc}
                    </button>
                </li>
            );
        })

        let status;
        if (winningData?.winner) {
            status = 'Winner: ' + winningData.winner;
        } else {
            status = 'Next player: ' + (this.state.isPlayerOneNext ? playerOneMark : playerTwoMark);
        }

        return (
            <div className="game">
                <div className="game-board">
                    <Board
                        squares={current.squares}
                        onClick={(i) => this.handleClick(i)}
                        winningSquares={winningData?.winningSquares}
                    />
                </div>
                <div className="game-info">
                    <div className="status">{status}</div>
                    <ol start="0">{moves}</ol>
                </div>
            </div>
        );
    }
}