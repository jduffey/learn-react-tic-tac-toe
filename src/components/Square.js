export function Square(props) {
    const bgColor = props.highlight ? "lightgreen" : "inherit";
    return (
        <button
            style={{ backgroundColor: bgColor }}
            className='square'
            onClick={props.onClick}
        >
            {props.value}
        </button>
    );
}