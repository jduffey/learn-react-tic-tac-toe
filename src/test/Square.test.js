import renderer from 'react-test-renderer';

import { Square } from "../components/Square";

describe("Square", () => {
    it("given isWinningSquare is true, sets background color to lightgreen", () => {
        const actual = renderer.create(
            <Square
                onClick={() => "foo"}
                playerMark={"bar"}
                isWinningSquare={true}
            />
        );

        expect(actual).toMatchSnapshot();
    });

    it("given isWinningSquare is false, sets background color to inherit", () => {
        const actual = renderer.create(
            <Square
                onClick={() => "foo"}
                playerMark={"bar"}
                isWinningSquare={false}
            />
        );

        expect(actual).toMatchSnapshot();
    });
})