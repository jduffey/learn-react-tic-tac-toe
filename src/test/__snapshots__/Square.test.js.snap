// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square given isWinningSquare is false, sets background color to inherit 1`] = `
<button
  className="square"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "inherit",
    }
  }
>
  bar
</button>
`;

exports[`Square given isWinningSquare is true, sets background color to lightgreen 1`] = `
<button
  className="square"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "lightgreen",
    }
  }
>
  bar
</button>
`;
